### PR TITLE
Release 8.0.5 PR

### LINKED ISSUE #242: Validation loss in tensorboard log remains 0
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

Using the simple command yolo to to test training:
If you then check tensorboard graphs, the loss of the validation remains a flat line at 0.
The yolov5 scripts did this correct.
Special option needed?

### Environment

_No response_

### Minimal Reproducible Example

_No response_

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### LINKED ISSUE #253: Problems with multiple gpu training
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

Traceback (most recent call last):
  File "/opt/conda/bin/yolo", line 8, in <module>
    sys.exit(cli())
  File "/opt/conda/lib/python3.8/site-packages/hydra/main.py", line 90, in decorated_main
    _run_hydra(
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/utils.py", line 394, in _run_hydra
    _run_app(
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/utils.py", line 457, in _run_app
    run_and_report(
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/utils.py", line 222, in run_and_report
    raise ex
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/utils.py", line 219, in run_and_report
    return func()
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/utils.py", line 458, in <lambda>
    lambda: hydra.run(
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/hydra.py", line 105, in run
    cfg = self.compose_config(
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/hydra.py", line 594, in compose_config
    cfg = self.config_loader.load_configuration(
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/config_loader_impl.py", line 142, in load_configuration
    return self._load_configuration_impl(
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/config_loader_impl.py", line 249, in _load_configuration_impl
    self.validate_sweep_overrides_legal(
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/config_loader_impl.py", line 79, in validate_sweep_overrides_legal
    raise ConfigCompositionException(msg)
hydra.errors.ConfigCompositionException: Ambiguous value for argument 'device=0,1,2,3'
1. To use it as a list, use key=[value1,value2]
2. To use it as string, quote the value: key=\'value1,value2\'
3. To sweep over it, add --multirun to your command line

### Environment

_No response_

### Minimal Reproducible Example

yolo task=detect mode=train model=yolov8m.pt data=fallset.yaml device=0,1,2,3 epochs=40 batch=240 lr0=0.001 v5loader=True

using [0,1,2,3] does not work either
4 GPUS RTX3090

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Laughing-q**
@AyushExel @glenn-jocher 
- make the following workflow work, now the val, predict and export following will use the trained model.
    ```python
  from ultralytics import YOLO

  # Load a model
  model = YOLO("yolov8n.yaml")  # build a new model from scratch
  model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)

  # Use the model
  results = model.train(data="coco128.yaml", epochs=3)  # train the model
  results = model.val()  # evaluate model performance on the validation set
  results = model("https://ultralytics.com/images/bus.jpg")  # predict on an image
  success = model.export(format="onnx")  # export a model to ONNX format
    ```
- update Docs, I think it's more clear, @AyushExel please take a look at this part, check if it makes sense to you.
image
- Set `return_output=False` as default in both python interface and cli cases. users already asked about it. #278

TODO: 
- [x] val nan loss issue
- [ ] decide whether to set `return_outputs=True` as default

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR introduces significant README updates and code improvements to the YOLOv8 model.

### 📊 Key Changes
- README file restructured for better readability and clarity.
- Codebase improvements with more Pythonic constructs and streamlined functionality for tasks like training, validation, prediction, and deployment.
- New `.md` files added for detailed documentation of tasks such as detection, segmentation, and classification.
- Updates in test cases to align with codebase changes and to ensure robustness.
- Enhanced exporter functionality to support model conversion to various formats with improved ease.
- Streamlined Python interfaces and CLI commands for improved user experience.

### 🎯 Purpose & Impact
- The update aims to make the model documentation more accessible and understandable, leading to a better user experience.
- By providing well-structured and detailed information on different tasks, users can more easily leverage YOLOv8 for their specific needs.
- Code enhancements contribute to a more efficient and reliable workflow, reducing potential errors and improving overall performance.
- The clear separation of tasks (detection, segmentation, classification) in documentation helps users to quickly find relevant instructions and guidelines, enhancing productivity.
- The test case improvements help in maintaining high code quality, ensuring that changes do not introduce regressions.
- Export features facilitate deploying models across various platforms, broadening the scope of applications and ease of integration into existing systems. 

🚀 These changes are set to improve the usability, scalability, and functionality of the YOLOv8 model, impacting a wide range of users from researchers to developers in the field of object detection and image analysis.

### CODE REVIEW

**Comment by Laughing-q** [comment]
@AyushExel also I don't see any obvious stuff about how to set args in our Docs. Maybe we can add some examples on the top of `Configuration` page to show users how to set args? 
something like this:
```python 
# for python interface
mode.train(imgsz=640, lr=0.01, copy_paste=0.5)
# for cli
yolo task=detect mode=train imgsz=640, lr=0.01, copy_paste=0.5
```

File: ultralytics/yolo/engine/model.py [hunk 1]
```diff
@@ -191,6 +191,9 @@ def train(self, **kwargs):
             self.trainer.model = self.trainer.get_model(weights=self.model if self.ckpt else None, cfg=self.model.yaml)
             self.model = self.trainer.model
         self.trainer.train()
+        # update model and configs after training
+        self.model, _ = attempt_load_one_weight(str(self.trainer.best))
+        self.overrides = self.model.args
```
Comment (AyushExel) [inline ultralytics/yolo/engine/model.py:196]: @Laughing-q what's the difference bw doing this and directly setting `self.model = self.trainer.model`? Are both same

**Comment by AyushExel** [comment]
> @AyushExel also I don't see any obvious stuff about how to set args in our Docs. Maybe we can add some examples on the top of `Configuration` page to show users how to set args? something like this:
> 
> ```python
> # for python interface
> mode.train(imgsz=640, lr=0.01, copy_paste=0.5)
> # for cli
> yolo task=detect mode=train imgsz=640, lr=0.01, copy_paste=0.5
> ```
> 
> EDIT: I saw some users had asked about how to set args yesterday if I remember this correctly.

yes you're rights we can add this example in the config section of docs

File: ultralytics/yolo/engine/model.py (hunk 1 above)
Comment (Laughing-q) [reply ultralytics/yolo/engine/model.py:196]: our export needs `pt_path` for file name which we give it to model in `attempt_load_one_weight`.

**Comment by Laughing-q** [comment]
@AyushExel If this PR makes sense to you, maybe you can take over this PR to update the Docs? add the config examples and review the page I updated.

File: ultralytics/yolo/engine/model.py (hunk 1 above)
Comment (AyushExel) [reply ultralytics/yolo/engine/model.py:196]: Okay got it

**Comment by AyushExel** [comment]
@Laughing-q okay yes. Let's just turn this into release 8.0.5 PR and all of us can commit here and link issues

**Comment by Laughing-q** [comment]
@AyushExel hey do you see the issues I linked? actually there's a bug in our loss calculation during evaluation. But I do remember there's a code version before we public the code, it worked correctly. So it could be some commits broke it, can you recall something about this?

**Comment by AyushExel** [comment]
@Laughing-q yes I recall a PR with title that went like "fix val issue" but I'm not sure if it was this issue. I'll look around

**Comment by Laughing-q** [comment]
@AyushExel oh thanks! I also found it and fixed it already. :)

**Comment by glenn-jocher** [comment]
@Laughing-q @AyushExel the reason I changed this line is that it comes after a training, and the export errors when the model is trained first. We need to update .train() to return the model, because I think right now the following lines only predict and export the original loaded model, not the trained model:

Screenshot 2023-01-12 at 11 21 31

```python
from ultralytics import YOLO

model = YOLO("yolov8n.pt")  # model A

# Use the model
results = model.train(data="coco128.yaml", epochs=3)  # train model B
results = model.val()  # val on model A or B?
results = model("https://ultralytics.com/images/bus.jpg")  # model A or B?
success = model.export(format="onnx")  # model A or B? Also error here, so replaced with YOLO("yolov8n.pt").export()
```

**Comment by Laughing-q** [comment]
@glenn-jocher oh I actually updated `self.model` to `self.train.model` in `train()` after the training finished, so this usage is going well.
But I didn't add `return model` in train().

**Comment by AyushExel** [comment]
@glenn-jocher @Laughing-q check slack..I've proposed a solution for handling predictions. Let me know if you guys agree and I'll start building the results class.

**Comment by Laughing-q** [comment]
@glenn-jocher I fixed a plots issue in [https://github.com/ultralytics/ultralytics/pull/292](https://github.com/ultralytics/ultralytics/pull/292). I assume you're working on this PR, so I opened another. please merge that one to this whenever you see this. :)

**Comment by AyushExel** [comment]
https://github.com/ultralytics/ultralytics/issues/280

**Comment by glenn-jocher** [comment]
@Laughing-q @AyushExel @izam-mohammed @yuewangg @ThibautLucas PR merged! Thank you for your contributions :)

**Pull request merged**